hail/python/hailtop/aiocloud/aiogoogle/client/storage_client.py

```diff
@@ -69,39 +69,53 @@ async def __anext__(self):
         raise StopAsyncIteration
 
 
+async def _cleanup_future(fut: asyncio.Future):
+    if not fut.done():
+        fut.cancel()
+    await asyncio.wait([fut])
+    if not fut.cancelled():
+        if exc := fut.exception():
+            raise exc
+
+
 class InsertObjectStream(WritableStream):
     def __init__(self, it: FeedableAsyncIterable[bytes], request_task: asyncio.Task[aiohttp.ClientResponse]):
         super().__init__()
         self._it = it
         self._request_task = request_task
         self._value = None
+        self._exit_stack = AsyncExitStack()
+
+        async def cleanup_request_task():
+            if not self._request_task.cancelled():
+                try:
+                    async with await self._request_task as response:
+                        self._value = await response.json()
+                except AttributeError as err:
+                    raise ValueError(repr(self._request_task)) from err
+            await _cleanup_future(self._request_task)
+
+        self._exit_stack.push_async_callback(cleanup_request_task)
 
     async def write(self, b):
         assert not self.closed
 
         fut = asyncio.ensure_future(self._it.feed(b))
-        try:
-            await asyncio.wait([fut, self._request_task], return_when=asyncio.FIRST_COMPLETED)
-            if fut.done() and not fut.cancelled():
-                if exc := fut.exception():
-                    raise exc
-                return len(b)
-            raise ValueError('request task finished early')
-        finally:
-            fut.cancel()
+        self._exit_stack.push_async_callback(_cleanup_future, fut)
+
+        await asyncio.wait([fut, self._request_task], return_when=asyncio.FIRST_COMPLETED)
+        if fut.done():
+            await fut
+            return len(b)
+        raise ValueError('request task finished early')
 
     async def _wait_closed(self):
-        fut = asyncio.ensure_future(self._it.stop())
         try:
+            fut = asyncio.ensure_future(self._it.stop())
+            self._exit_stack.push_async_callback(_cleanup_future, fut)
             await asyncio.wait([fut, self._request_task], return_when=asyncio.FIRST_COMPLETED)
-            async with await self._request_task as resp:
-                self._value = await resp.json()
         finally:
-            if fut.done() and not fut.cancelled():
-                if exc := fut.exception():
-                    raise exc
-            else:
-                fut.cancel()
+            await self._exit_stack.aclose()
```

Comment:
Can you double check the exit stack runs in the order you intended it to run in? I know we messed that up with Batch before.

Reply:
Yes, it is in the correct order. Above, I push onto the stack whenever I create a resource, so the resources are on the stack in their order of creation. They are then popped one by one from the stack so the most recently created is closed first, as it it would be if you had nested try-finallies.

The bug in Batch was because we did:
```python3
def close():
    async with ExitStack() as stack:
        stack.push(resource_we_wanted_to_close_first)
        stack.push(resource_we_wanted_to_close_second)
```
When the ExitStack runs `__exit__`, it pops the stack and sees "resource_we_wanted_to_close_second"  first and closes that first. 

The root cause of the issue in Batch was trying to use exit stacks only in the close method. They're supposed to be used the way I've used them here: you create them before any resources, push resources onto the stack as you create them, and then close them when all the resources are down.

Any time you see the _creation_ of an ExitStack in a close or cleanup method, that's almost certainly a bug.